### PR TITLE
Fast Test - Full compile on proto changes

### PR DESCRIPTION
**Pull request by mildagle**
# Checklist

- [ ] I have filed an issue about this change and discussed potential changes with the maintainers.
- [ ] I have received the approval from the maintainers to make this change.
- [ ] This is not a stylistic, refactoring, or cleanup change.

Please note that the maintainers will not be reviewing this change until all checkboxes are ticked. See 
the [Contributions](https://github.com/bazelbuild/intellij#contributions) section in the README for more 
details.

# Discussion thread for this change

Issue number: `<please reference the issue number or url here>`

# Description of this change

### CODE REVIEW

File: java/src/com/google/idea/blaze/java/fastbuild/FastBuildChangedFilesService.java [hunk 1]
```diff
@@ -74,7 +74,8 @@ static FastBuildChangedFilesService getInstance(Project project) {
 
   private static final Logger logger = Logger.getInstance(FastBuildChangedFilesService.class);
 
-  @VisibleForTesting static final int MAX_FILES_TO_COLLECT = 30;
```
Comment (ittaiz) [inline java/src/com/google/idea/blaze/java/fastbuild/FastBuildChangedFilesService.java:77]: I would revert to minimize potential conflicts later on

File: java/src/com/google/idea/blaze/java/fastbuild/FastBuildChangedFilesService.java [hunk 2]
```diff
@@ -156,11 +157,21 @@ synchronized ChangedSources getAndResetChangedSources(Label label) {
 
     switch (data.state) {
       case WAITING_FOR_SOURCES:
+        if (!data.changedNonCompilableSources.isEmpty()) {
```
Comment (ittaiz) [inline java/src/com/google/idea/blaze/java/fastbuild/FastBuildChangedFilesService.java:160]: Why does this happen only in waiting for sources and not in collecting?

File: java/src/com/google/idea/blaze/java/fastbuild/FastBuildChangedFilesService.java [hunk 3]
```diff
@@ -356,5 +399,17 @@ void updateChangedSources(Set<File> changedFiles) {
         state = State.TOO_MANY_CHANGES;
       }
     }
+
+    void checkForNonCompilableChanges(Set<File> changedNonCompileFiles) {
```
Comment (ittaiz) [inline java/src/com/google/idea/blaze/java/fastbuild/FastBuildChangedFilesService.java:403]: why separate this logic from updatechangedsources? Should this be in the same method?

File: java/src/com/google/idea/blaze/java/fastbuild/FastBuildChangedFilesService.java [hunk 4]
```diff
@@ -356,5 +399,17 @@ void updateChangedSources(Set<File> changedFiles) {
...
+      if (state.equals(State.WAITING_FOR_SOURCES)) {
```
Comment (ittaiz) [inline java/src/com/google/idea/blaze/java/fastbuild/FastBuildChangedFilesService.java:404]: Why only in WAITING_FOR_SOURCES? What about Collecting?

File: java/src/com/google/idea/blaze/java/fastbuild/FastBuildChangedFilesService.java (hunk 2 above)
Comment (mildagle) [reply java/src/com/google/idea/blaze/java/fastbuild/FastBuildChangedFilesService.java:160]: This is due to the specific case, where proto files were edited, but the state is still "WAITING_FOR_SOURCES" which means, that the edited files could not be compared to sources, and full compile happens "just in case". If the state is COLLECTING, it means, that the edited files are being checked against the sources, so there weren't any changed proto files that were in the sources. If there were changed files that were in the sources , the state would be set to NON_COMPILABLE_CHANGES, and full compile would happen due to that. 

If not for this specific case, the code would be much more simple, as there would not be need for data.changedNonCompilableSources property at all :/

File: java/src/com/google/idea/blaze/java/fastbuild/FastBuildChangedFilesService.java (hunk 4 above)
Comment (mildagle) [reply java/src/com/google/idea/blaze/java/fastbuild/FastBuildChangedFilesService.java:404]: As mentioned in the other comment, it's for the case, were proto file changes were made, but the state is WAITING_FOR_SOURCES, so they cannot be compared to actual sources, and if getAndResetChangedSources is called before the sources are collected, full compile will have to happen. If the sources were collected, the files in changedNonCompilableSources property will be checked against them, and the state will be set to NON_COMPILABLE_CHANGES. As well as if the state was COLLECTING, and any proto files from the sources were changed, the NON_COMPILABLE_CHANGES would be set too.

File: java/src/com/google/idea/blaze/java/fastbuild/FastBuildChangedFilesService.java (hunk 3 above)
Comment (mildagle) [reply java/src/com/google/idea/blaze/java/fastbuild/FastBuildChangedFilesService.java:403]: Because they are not doing the same thing. The job of updateChangedSources is to take the changed files, filter the relevant ones from the sources, and adding them to changedSources property, which is the property, that collects the files that get compiled when getAndResetChangedSources. Setting the MAX_FILES_TO_COLLECT state is more like a side effect for this method, and not the main purpose.  On the other hand, the main purpose of checkForNonCompilableChanges is to check incoming proto files and if any of them is in sources, setting the State. It does not collect them (except for the WAITING_FOR_SOURCES case, to do the full compile if  getAndResetChangedSources gets called before the sources are set) 

Though now rethinking everything, maybe not adding the additional state, collecting the non compilable changes and checking for them in WAITING_FOR_SOURCES/COLLECTING  at getAndResetChangedSources  is better for the future, in case we do want to do something with these files later on

**Comment by ittaiz** [comment]
I'm leaving a general comment because I think most of my comments (and your answers) were indeed around the same dilemma and that is the addition of the additional state in the state machine.
I think the state machine before your change is already pretty complex and I'd rather not add another state option there.
Especially since I'm not sure what's the value of it.
An alternative can be to have this logic somewhere else and then call `ChangedSources.fullCompile()` where applicable.

One potential caveat to the approach I'm suggesting is that irrelevant files get to incremental compilers
```
performIncrementalCompilation(
          context, label, existingBuildState, changedSources.changedSources());
```
OTOH like you mention above maybe it's ok and we "just" change incremental compilers to filter out files they can't handle.
And then (again like you mention) if someone decides they want to have proto incremental compilation they can. 
Don't know how important that is.

Bottom line- I'm torn. I feel the current change is a bit unclear and at the same time my alternative also adds ambiguity.
Maybe your current change is the right direction but it requires another method, a bit more naming and maybe a comment or two? on the why.

File: java/src/com/google/idea/blaze/java/fastbuild/FastBuildChangedFilesService.java (hunk 3 above)
Comment (ittaiz) [reply java/src/com/google/idea/blaze/java/fastbuild/FastBuildChangedFilesService.java:403]: > The job of updateChangedSources is to take the changed files, filter the relevant ones from the sources, and adding them to changedSources property, which is the property, that collects the files that get compiled when getAndResetChangedSources.

Why does this not apply to protos? Even if we stay with this approach then it's worthwhile explaining why

File: java/src/com/google/idea/blaze/java/fastbuild/FastBuildChangedFilesService.java (hunk 4 above)
Comment (ittaiz) [reply java/src/com/google/idea/blaze/java/fastbuild/FastBuildChangedFilesService.java:404]: I think this isn't a problem. This can happen also for scala and java files, right? That they're called when WAITING_FOR_SOURCES is the state. What's the problem?

Comment (mildagle) [reply java/src/com/google/idea/blaze/java/fastbuild/FastBuildChangedFilesService.java:404]: It is not a problem. It's just the part, that makes the implementation less straight forward, as you can see, the WAITING_FOR_SOURCES state jumps out in multiple methods.

File: java/src/com/google/idea/blaze/java/fastbuild/FastBuildChangedFilesService.java (hunk 3 above)
Comment (mildagle) [reply java/src/com/google/idea/blaze/java/fastbuild/FastBuildChangedFilesService.java:403]: Because with Java/Scala files, the changed files need to be collected, as they are the files that get compiled. In case of proto, you don't really need to track the files, you need to only know, that any changed proto files exist

**Comment by mildagle** [comment]
I don't want to just let the compilers to filter out files, I do agree, that it adds ambiguity.

I actually think, that my current implementation is mostly unclear, because the behaviour for uncompilable changed sources is very similar, but also too different, compared to compilable sources.

I implemented alternative PR, where the tracking of non-compilable sources is still explicit, but it's done the same way as compilable sources, only in the getAndResetChangedSources it's checked, if any exist.  I think it makes it much more clear, and actually does not mean any files need to be tracked after a single proto file change is detected, which I tried to prevent with the current implementation. So I prefer the new PR - https://github.com/wix-playground/intellij/pull/77
